include/wil/registry.h

```diff
@@ -17,16 +17,2951 @@
+        HRESULT get_value_nothrow(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, _Out_ T* return_value) WI_NOEXCEPT
+        {
+            const reg_view_details::reg_view_nothrow regview{ key };
+            return regview.get_value<T>(subkey, value_name, *return_value);
+        }
+
+        /**
+         * \brief Reads a value under a specified key, the registry type based off the templated type passed as data
+         * \tparam T The type capturing the data being set
+         * \param key An opened registry key, or fixed registry key as the base key, from which to append the path
+         * \param value_name A string specifying the name of the registry value to read from
+         *        can be nullptr to read from the unnamed default registry value
+         * \param[out] return_value A pointer-to-T receiving the value read from the registry
+         * \return HRESULT error code indicating success or failure (does not throw C++ exceptions)
+         */
+        template<typename T,
+            std::enable_if_t<!std::is_same_v<T, wchar_t>>* = nullptr>
+        HRESULT get_value_nothrow(HKEY key, _In_opt_ PCWSTR value_name, _Out_ T* return_value) WI_NOEXCEPT
+        {
+            return ::wil::reg::get_value_nothrow(key, nullptr, value_name, return_value);
+        }
+
+        /**
+         * \brief Reads a REG_SZ value under a specified key
+         * \tparam Length The length of the WCHAR array passed as an OUT parameter
+         * \param key An opened registry key, or fixed registry key as the base key, from which to append the path
+         * \param subkey A string to append to the HKEY to attempt to read from
+         *        can be nullptr if not needed
+         * \param value_name A string specifying the name of the registry value to read from
+         *        can be nullptr to read from the unnamed default registry value
+         * \param[out] return_value A WCHAR array receiving the value read from the registry
+         *             will write to the WCHAR array the string value read from the registry, guaranteeing null-termination
+         * \param[out] requiredBytes An optional pointer to a DWORD to receive the required bytes of the string in the registry
+         * \return HRESULT error code indicating success or failure (does not throw C++ exceptions)
+         */
+        template <size_t Length>
+        HRESULT get_value_string_nothrow(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, WCHAR(&return_value)[Length], _Out_opt_ DWORD* requiredBytes = nullptr) WI_NOEXCEPT
+        {
+            const reg_view_details::reg_view_nothrow regview{ key };
+            return regview.get_value_char_array(subkey, value_name, return_value, REG_SZ, requiredBytes);
+        }
+        template <size_t Length>
+        HRESULT get_value_string_nothrow(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, WCHAR(&return_value)[Length], _Out_opt_ uint32_t* requiredBytes) WI_NOEXCEPT

```

Comment:
it's because we are passing by pointer. it covered more types. and I *really* want to avoid forcing callers to play static-cast-integral-types games with this interface :)

Reply:
I get that, but how much more likely are folks to have a `uint32_t` versus a `DWORD`? `size_t` on 64bit will be `uint64_t`, and `int` is its own thing.

`DWORD` and `uint32_t` are basically equivalent types (guaranteed to be "at least" 32-bit vs 32-bit exactly), and I understand the rationale for using the C++ builtin types, but I'm not convinced callers are more likely to have one versus the other.

I don't feel strongly enough about this that it's blocking, but if it were up to me, I'd remove the `uint32_t` overloads for simplicity's sake. I don't think they make the API a ton more ergonomic in practice, and they add enough complexity to be frustrating (default values in overloads, eg).

Thoughts?

Reply:
This happened when I was using the API with a variety of integer types. All 32-bit. And I hate interfaces that make me play static-cast-to-the-right-type to use it.

Reply:
I fought with this for a while to make this more generic. I'll just keep DWORD for now. I can reasses this later and add interfaces as needed without breaking changes.